Make the docker log processor use the global DockerUtil rather than its own docker client

L3n41c opened the pull request:
### What does this PR do?

This change removes the docker client of the log package so that there's a single docker client.

### Motivation

The package `pkg/logs/input/docker` used to have its own docker client.
This could create some unwanted behaviour in some edge cases.

One example is if the docker client used by the docker auto-discovery listener is successfully connected to the docker daemon, it can start sending "new docker source" and "new docker service" events to the log package.
If the docker client used by the log package is not connected to the docker daemon, the "new docker source" and "new docker service" will not be properly processed.

### Additional Notes

This is followup action of #4839

L3n41c committed:
Make the docker log processor use the global DockerUtil rather than its own docker client

The package `pkg/logs/input/docker` used to have its own docker client.
This could create some unwanted behavior in some edge cases.

One example is if the docker client used by the docker auto-discovery listener
is successfully connected to the docker daemon, it can start sending "new
docker source" and "new docker service" events to the log package.
If the docker client used by the log package is not connected to the docker
daemon, the "new docker source" and "new docker service" will not be properly
processed.

This change removes the docker client of the log package so that there's a
single docker client.
This will fix the behavior in the above described edge case.

L3n41c committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into lenaic/remove_log_docker_client

ajacquemot commented on pkg/logs/input/docker/client.go:
```diff
@@ -1,82 +0,0 @@
-// Supported versions of the Docker API
-const (
-	minVersion = "1.18"
-	maxVersion = "1.25"
```
Are we sure that nothing will break without those two checks ? e.g. what happens if a user has a version prior to `1.18` with the new change ?

L3n41c replied:
Most of the `DockerUtil` services that are now used in the log agent code were already existing and were already used by some other parts of the agent.
The only function that I added to `DockerUtil` to support the use cases of the log agent is `ContainerLogs`:
https://github.com/DataDog/datadog-agent/blob/386a22ab3b0efdfc290d6667160d2b5625960736/pkg/util/docker/containers.go#L162-L164
And the API used to retrieve the logs is the same (same URL and same parameters) between the [oldest documented API version (1.18)](https://docs.docker.com/engine/api/v1.18/#get-container-logs) and the [newest documented API version (1.40)](https://docs.docker.com/engine/api/v1.40/#operation/ContainerLogs).

[The docker website is not documenting API versions prior to `1.18` anymore.](https://docs.docker.com/engine/api/version-history/)

The docker client that we vendor doesn’t document versions prior to `1.18` either:
```
~/devel/DataDog/datadog-agent $ ls vendor/github.com/docker/docker/docs/api
v1.18.md  v1.19.md  v1.20.md  v1.21.md  v1.22.md  v1.23.md  v1.24.md  version-history.md
```

Here is the PR that introduced this check: #1164.

Do you remember if there was an issue with docker API version prior to `1.18` or if we added this check just to avoid untested setup because we didn’t validate the agent with old docker daemon ?

Anyhow, before this change, if the agent tried to talk to a docker daemon which doesn’t support API 1.18 or above, we would fail the instantiation of the docker client. But it wouldn’t prevent the log agent from receiving new services and sources notification because the AD was already using `GetDockerUtil`.
With this change, the instantiation of the client would succeed. If the API version doesn’t support log retrieval, that’s the `ContainerLogs()` function that will return an error.

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -151,12 +97,12 @@ func (l *Launcher) run() {
 		case service := <-l.addedServices:
 			// detected a new container running on the host,
-			cli, err := l.getCli()
+			dockerutil, err := dockerutil.GetDockerUtil()
```
do we want to keep a reference to dockerutil instead of calling it for each new service ? What can go wrong ?

L3n41c replied:
I’ve introduced the `l.getCli()` earlier in order to be more resilient to situations where the agent is started before it is able to connect to the docker daemon.
In those situations, we might get a non-`nil` error and a `nil` client.
So, we would need to cache the client when it succeeds to connect to the docker daemon and to retry later when it fails.
This was the purpose of the `retrier` that there was inside the `getCli()`.
That retrier is now inside `GetDockerUtil()`: https://github.com/DataDog/datadog-agent/blob/4cc1fe3ca3f14aa752b63f0c044c01a9029a86bc/pkg/util/docker/global.go#L40-L44

The caching logic is in fact inside `GetDockerUtil()`.

ajacquemot replied:
sounds good, thanks for the explanation

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -248,12 +194,12 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
 	// overridenSource == source if the containerCollectAll option is not activated or the container has AD labels
 	overridenSource := l.overrideSource(container, source)
-	cli, err := l.getCli()
+	dockerutil, err := dockerutil.GetDockerUtil()
```
Same question here

ajacquemot reviewed (commented):
Only dumb questions and maybe one which is important to answer I think: what happens with the client for versions of docker prior to `1.18`, if I remember correctly it's the from where the endpoint for collections was released.

gaetan-deputier commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -58,14 +57,14 @@ type Tailer struct {
 		source:             source,
 		tagProvider:        tag.NewProvider(dockerutil.ContainerIDToTaggerEntityName(containerID)),
-		cli:                cli,
+		dockerutil:         dockeruti,
```
not `dockeruti` -> `dockerutil`

L3n41c replied:
I put back the old name `cli`. In this context, we cannot name the variable `dockerutil` because it would hide the `dockerutil` package.

L3n41c committed:
Rename `dockeruti` to `cli`

`dockerutil` cannot be used as the name of the variable would hide the name of the package
